src/ca/idexchange.rs

```diff
@@ -0,0 +1,1199 @@
+/// following - taken from the RELAX NG schema in RFC 8183:
+/// 
+/// handle  = xsd:string { maxLength="255" pattern="[\-_A-Za-z0-9/]*" }
+#[derive(Clone, Debug, Deserialize, Eq, Hash, PartialEq)]
+#[serde(try_from = "String")]
+pub struct Handle {
+    name: Arc<str>,
+}
+
+impl Handle {
+    pub fn as_str(&self) -> &str {
+        self.as_ref()
+    }
+
+    /// We replace "/" with "+" and "\" with "=" to make file system
+    /// safe names.
+    pub fn to_path_buf(&self) -> PathBuf {
+        let s = self.to_string();
+        let s = s.replace('/', "+");
+        let s = s.replace('\\', "=");
+        PathBuf::from(s)
+    }
+
+    fn verify_name(s: &str) -> Result<(), InvalidHandle> {
+        if s.bytes()
+            .all(|b| {
+                b.is_ascii_alphanumeric() || b == b'-' || b == b'_' ||
+                b == b'/' || b == b'\\'
+            })
+            && !s.is_empty()
+            && s.len() < 256
+        {
+            Ok(())
+        } else {
+            Err(InvalidHandle)
+        }
+    }
+}
+
+impl TryFrom<&PathBuf> for Handle {
+    type Error = InvalidHandle;
+
+    fn try_from(path: &PathBuf) -> Result<Self, Self::Error> {
+        if let Some(path) = path.file_name() {
+            let s = path.to_string_lossy().to_string();
+            let s = s.replace('+', "/");
+            let s = s.replace('=', "\\");
+            Self::from_str(&s)
+        } else {
+            Err(InvalidHandle)
+        }
+    }
+}
+
+impl FromStr for Handle {
+    type Err = InvalidHandle;
+
+    /// Accepted pattern: [-_A-Za-z0-9/]{1,255}
+    /// See Appendix A of RFC8183.
+    ///
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        Self::verify_name(s)?;
+        Ok(Handle { name: s.into() })
+    }
+}
+
+impl TryFrom<String> for Handle {
+    type Error = InvalidHandle;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Self::verify_name(&value)?;
+        Ok(Handle { name: value.into() })
+    }
+}
+
+impl AsRef<str> for Handle {
+    fn as_ref(&self) -> &str {
+        &self.name
+    }
+}
+
+impl AsRef<[u8]> for Handle {
+    fn as_ref(&self) -> &[u8] {
+        self.name.as_bytes()
+    }
+}
+
+impl fmt::Display for Handle {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.as_str())
+    }
+}
+
+impl Serialize for Handle {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.as_str().serialize(serializer)
+    }
+}
+
+#[derive(Debug)]
+pub struct InvalidHandle;
+
+impl fmt::Display for InvalidHandle {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Handle MUST have pattern: [-_A-Za-z0-9/]{{1,255}}")
+    }
+}
+
+
+//------------ ServiceUri ----------------------------------------------------
+
+/// The service URI where a child or publisher needs to send its
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum ServiceUri {
```

Comment:
Improved the doc comment.

I like having `ServiceUri` here because this is also a named concept in RFC 8183. We could have `rpki::uri::Http` instead of `String`, but in this context it does not really matter as we are always just using the `&str` of either variant when posting protocol messages.

So, in short I propose to postpone doing that work. Because, if we would have an `rpki::uri::Http` then it would most likely need to have more features.